contracts/core/MarginRequirements.sol

```diff
@@ -0,0 +1,185 @@
+
+    /**
+     * @notice sets the initial margin %
+     * @dev can only be called by owner
+     * @param _underlying underlying asset address
+     * @param _collateralAsset collateral asset address
+     * @param _isPut option type the vault is selling
+     * @param _account account address
+     * @param _initialMargin initial margin percentage (eg. 10% = 1000)
+     */
+    function setInitialMargin(
+        address _underlying,
+        address _collateralAsset,
+        bool _isPut,
+        address _account,
+        uint256 _initialMargin
+    ) external onlyOwner {
+        require(
+            _initialMargin > 0 && _initialMargin <= 100 * 10**2,
+            "MarginRequirements: initial margin cannot be 0 or higher than 100%"
+        );
+        require(_underlying != address(0), "MarginRequirements: invalid underlying");
+        require(_collateralAsset != address(0), "MarginRequirements: invalid collateral");
+        require(_account != address(0), "MarginRequirements: invalid account");
+
+        initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][_account] = _initialMargin;
+    }
+
+    /**
+     * @notice sets the maintenance margin absolute amount
+     * @dev can only be called by keeper
+     * @param _vaultID id of the vault
+     * @param _maintenanceMargin maintenance margin absolute amount
+     */
+    function setMaintenanceMargin(uint256 _vaultID, uint256 _maintenanceMargin) external onlyKeeper {
+        require(_maintenanceMargin > 0, "MarginRequirements: initial margin cannot be 0");
+
+        maintenanceMargin[_vaultID] = _maintenanceMargin;
+    }
+
+    /************************************************
+     *  MARGIN OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice checks if there is enough collateral to mint the desired amount of otokens
+     * @param _account account address
+     * @param _notional order notional amount
+     * @param _underlying underlying asset address
+     * @param _isPut option type the vault is selling
+     * @param _collateralAsset collateral asset address
+     * @param _collateralAmount collateral amount
+     * @return boolean value stating whether there is enough collateral to mint
+     */
+    function checkMintCollateral(
+        address _account,
+        uint256 _notional,
+        address _underlying,
+        bool _isPut,
+        uint256 _collateralAmount,
+        address _collateralAsset
+    ) external view returns (bool) {
+        uint256 collateralDecimals = uint256(ERC20Interface(_collateralAsset).decimals());
+
+        uint256 initialMarginRequired = initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][
+            _account
+        ];
+
+        return
+            _notional.mul(initialMarginRequired).mul(10**collateralDecimals).mul(1e8) <
+            _collateralAmount.mul(oracle.getPrice(_collateralAsset)).mul(100e2).mul(1e6);
+    }
+
```

Comment:
recommend adding constants to margin requirements rather than magic numbers:

1. MAX_INITIAL_MARGIN = 100 * 10**2
2. NOTIONAL_DECIMALS = 6
3. also use oracle.decimals() over 10**8

We can then use these in checkMintCollateral, checkWithdrawCollateral, setInitialMargin. Other magic numbers can also put as constants
5.

Reply:
Went double-checking - I was under the impression and it seems that oracle.decimals() is fixed value of 8.

The steps the price goes through to get to the oracle seem to be:

1. Originates in `Chainlink Aggregator`

4. Then it is called by `Chainlink Pricer` - Which has a function `_scaleToBase` which transforms the value coming from the `Chainlink Aggregator` regardless of decimals into a number with 8 decimals

- For reference 1: Risk-fork `ChainlinkPricer.sol` for reference `_scaleToBase` below:
<img width="300" alt="image" src="https://user-images.githubusercontent.com/99089658/226092845-5a5a4fd0-e8d4-46b8-9492-bc06ace2883b.png">

- For reference 2: ETH Mainnet deployment: link [here](https://etherscan.io/address/0x4A355AEcB08A381281cfFb801799faAF025a7EfB#code) - see below:
<img width="620" alt="image" src="https://user-images.githubusercontent.com/99089658/226092914-27c342fc-6e5e-4623-b7f1-c244efd161ed.png">


3. This transformed price with 8 decimals is then what we get as an output on `Oracle`

Might I be missing something?

PS: committed a version with `MAX_INITIAL_MARGIN`, `NOTIONAL_DECIMALS`, and `ORACLE_DECIMALS` as constants here that is pending confirmation of the above: 


https://github.com/ribbon-finance/GammaProtocol-OTC/pull/2/commits/5fa8d09ee30b17b96e7c280b220442cdc28acb7a

Reply:
we can hardcode 8 decimals for chainlink pricer, but for some non-mainnet assets we will need to have a [manual pricer](https://github.com/opynfinance/GammaProtocol/pull/463/files). Can you create a scale to base for those as well? We will also need to build a script so that we can't fatfinger it entering the pricer.